### PR TITLE
fix: allow editing share type posts (squad only)

### PR DESCRIPTION
**Pull request by DragosIorgulescu**
## Dependencies

- https://github.com/dailydotdev/daily-api/pull/1461

## Changes

- updates the `sharePost` mutation usage to allow passing a `postID` and behave like an upsert operation
- use the existing `ShareLink` component to render the correct form when editing a `share` post type

## Preview

edit-share-post

## Events

_N / A_

### On those affected packages:
- [x] Have you done sanity checks in the webapp?
- [x] Have you done sanity checks in the extension?
- [x] Does this not break anything in companion?

### Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [ ] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

WT-1801 #done

### CODE REVIEW

File: packages/shared/src/components/post/write/WriteLinkPreview.tsx [hunk 1]
```diff
@@ -48,7 +48,7 @@ export function WriteLinkPreview({
         <WritePreviewContent className={isMinimized && '!py-2 !px-3'}>
           <div className="flex flex-col flex-1 typo-footnote">
             <span className="font-bold line-clamp-2">{preview.title}</span>
-            {preview.source &&
+            {preview.source?.id !== 'unknown' &&
```
Comment (DragosIorgulescu) [inline packages/shared/src/components/post/write/WriteLinkPreview.tsx:51]: This does not feel right, but based on my understanding of how these are stored in the db, it's the only way to determine if this is a URL preview of an external page, or a preview of an internal post

**Comment by DragosIorgulescu** [comment]
I still have some UI cleanup to do, but overall I tried to be as uninstrusive as possible with implementing this.

File: packages/shared/src/components/post/write/WriteLinkPreview.tsx (hunk 1 above)
Comment (rebelchris) [reply packages/shared/src/components/post/write/WriteLinkPreview.tsx:51]: Yep it's a set constraint at the moment :)

File: packages/webapp/pages/posts/[id]/edit.tsx [hunk 2]
```diff
@@ -104,7 +105,19 @@ function EditPost(): ReactElement {
         isForbidden={!isVerified || !squad || !canEdit}
       >
         <WritePostHeader isEdit />
-        <WriteFreeformContent className="py-6 px-4" />
+
+        {post?.type === PostType.Share ? (
```
Comment (rebelchris) [inline packages/webapp/pages/posts/[id]/edit.tsx:109]: I think ideally we use a completely different mutation for editing actually.
Like we do for EditPost it's a better failsafe way.

Not 100% sure how it's tackled in the freeform one now, but maybe we can look at that logic and re-apply?

**Review by rebelchris — COMMENTED** [review]
Think we can make it safer by applying seperate calls.

File: packages/shared/src/hooks/squads/usePostToSquad.ts [hunk 3]
```diff
@@ -149,10 +171,33 @@ export const usePostToSquad = ({
     [preview, displayToast, onSubmitLink, onPost, isPosting],
   );
 
+  const onUpdatePost = useCallback(
+    (
+      e: BaseSyntheticEvent,
+      postId: Post['id'],
+      sourceId: Squad['id'],
+      commentary: string,
```
Comment (rebelchris) [inline packages/shared/src/hooks/squads/usePostToSquad.ts:179]: This is basically double-defined in the interface and here, can't we make it 1 type to share across both instances?

**Review by rebelchris — COMMENTED** [review]
One small suggestion.
After API is resolved, I'd like to test it out locally before we merge this, so holding my approval for that 🙏

File: packages/shared/src/hooks/squads/usePostToSquad.ts (hunk 3 above)
Comment (DragosIorgulescu) [reply packages/shared/src/hooks/squads/usePostToSquad.ts]: good point, cleaned it up 🙏

File: packages/shared/src/hooks/squads/usePostToSquad.ts [hunk 4]
```diff
@@ -149,10 +171,28 @@ export const usePostToSquad = ({
...
 
+  const onUpdatePost = useCallback<UsePostToSquad['onUpdatePost']>(
+    (e, postId, sourceId, commentary) => {
+      e.preventDefault();
+
+      if (isUpdating) {
+        return null;
+      }
+
+      return updatePost({
```
Comment (rebelchris) [inline packages/shared/src/hooks/squads/usePostToSquad.ts:182]: For some reason I get the generic "your post has been shared to squad" maybe we need a different callback here?
(also not sure it fires the same tracking perhaps?)

**Review by rebelchris — CHANGES_REQUESTED** [review]
When placing a new shared post (no edit) it fails now:
Screenshot 2023-10-13 at 16 57 35

File: packages/shared/src/hooks/squads/usePostToSquad.ts [hunk 5]
```diff
@@ -77,8 +84,12 @@ export const usePostToSquad = ({
       },
     });
 
-  const onSharedPostSuccessfully = async () => {
-    displayToast('This post has been shared to your Squad');
+  const onSharedPostSuccessfully = async (update = false) => {
```
Comment (rebelchris) [inline packages/shared/src/hooks/squads/usePostToSquad.ts:87]: For now no action needed, but think moving forward let's prefer object ways?

**Review by rebelchris — APPROVED** [review]
Nice tested and validated 🤘

File: packages/shared/src/graphql/squads.ts [hunk 6]
```diff
@@ -187,6 +187,14 @@ export const ADD_POST_TO_SQUAD_MUTATION = gql`
   }
 `;
 
+export const UPDATE_SQUAD_POST_MUTATION = gql`
+  mutation UpdateSquadPost($id: ID!, $sourceId: ID!, $commentary: String) {
```
Comment (sshanzel) [inline packages/shared/src/graphql/squads.ts:191]: Why do we need the `sourceId` as part of the parameters? Should the ID alone make it unique enough?

**Review by sshanzel — APPROVED** [review]
Looks great! Just had some non blocking comments 🚀

File: packages/shared/src/graphql/squads.ts (hunk 6 above)
Comment (DragosIorgulescu) [reply packages/shared/src/graphql/squads.ts:191]: not really needed, but does not hurt to have it either, my reasoning is to query based on both source & id to ensure we are not accidentally accessing the wrong record.

Comment (rebelchris) [reply packages/shared/src/graphql/squads.ts:191]: I would also just stick with ID it's uniform with what we already use for other pieces and reduces overhead I think.

**Pull request merged**